### ISSUE TITLE
Support AWS provider version 4.0.0

### ISSUE THREAD

**Issue by HorizonNet** [bridgecrewio/checkov#2399]
**Describe the issue**
The latest AWS provider version 4.0.0 introduces multiple breaking changes, especially regarding the `aws_s3_bucket` resource

**Examples**
The following snippet is based on version 4.0.0 and will make the following checks fail

* aws-s3-enable-bucket-encryption
* aws-s3-enable-versioning
* aws-s3-encryption-customer-key

```terraform
resource "aws_s3_bucket" "example" {
  bucket = "example"
}

resource "aws_s3_bucket_acl" "example" {
  bucket = aws_s3_bucket.example.id
  acl    = "private"
}

resource "aws_s3_bucket_server_side_encryption_configuration" "example" {
  bucket = aws_s3_bucket.example.id

  rule {
    apply_server_side_encryption_by_default {
      kms_master_key_id = aws_kms_key.example.arn
      sse_algorithm     = "aws:kms"
    }
  }
}

resource "aws_s3_bucket_versioning" "example" {
  bucket = aws_s3_bucket.example.id

  versioning_configuration {
    status = "Enabled"
  }
}

resource "aws_kms_key" "example" {
  description         = "example"
  enable_key_rotation = true
  multi_region        = false
}
```

**Version (please complete the following information):**
 - Checkov Version 2.0.823 (probably more versions)

**Additional context**
See the [migration guide](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/guides/version-4-upgrade).

**Comment by gruebel** [comment]
Hi @HorizonNet thanks for creating the issue 💪 

We will use it to track our effort to complain with the recent changes in the new provider version.

**Comment by babyhuey** [comment]
Does closing this issue mean that the issue is fixed? I'm still seeing failures due to my changes to support 4.x branch

**Comment by MrHash** [comment]
+1 specifically it's not recognizing the CMK encryption config on some buckets for me

**Comment by gruebel** [comment]
@MrHash I checked and you are right the CMK check was not ported to the new syntax will add a fix. @babyhuey any specific use cases, which don't work for you?

**Comment by babyhuey** [comment]
So the biggest one I've found is CKV_AWS_19. It is saying to "Ensure all data stored in the S3 bucket is securely encrypted at rest", but I definitely have a aws_s3_bucket_server_side_encryption_configuration defined for the buckets

Also Check: CKV_AWS_145: "Ensure that S3 buckets are encrypted with KMS by default"
is also failing, but I definitely have aws_s3_bucket_logging for those buckets defined. 

Can I no longer do a `checkov -d .` anymore and I'll have to use the json out file?

**Comment by gruebel** [comment]
@babyhuey nope it should still work with `-d .` that was our goal at least 😄 

I already created a PR to fix `CKV_AWS_145` check. So, maybe tomorrow it will be already merged, depends on my colleagues.

Can you provide an example for `CKV_AWS_19`, which result in a failure for you, so we can better investigate and understand what is still missing.

**Comment by babyhuey** [comment]
Sure. I'll give you a live example
```
Check: CKV_AWS_19: "Ensure all data stored in the S3 bucket is securely encrypted at rest"
        FAILED for resource: aws_s3_bucket.zappa-uap-us-west-1
        File: /s3.tf:2646-2649
        Guide: https://docs.bridgecrew.io/docs/s3_14-data-encrypted-at-rest

                2646 | resource "aws_s3_bucket" "zappa-uap-us-west-1" {
                2647 |   bucket   = "zappa-uap-us-west-1"
                2648 |   provider = aws.us-west-1
                2649 | }
```
But in my code I have
```
resource "aws_s3_bucket" "zappa-uap-us-west-1" {
  bucket   = "zappa-uap-us-west-1"
  provider = aws.us-west-1
}

resource "aws_s3_bucket_server_side_encryption_configuration" "zappa-uap-us-west-1" {
  bucket   = aws_s3_bucket.zappa-uap-us-west-1.id
  provider = aws.us-west-1
  rule {
    apply_server_side_encryption_by_default {
      sse_algorithm = "AES256"
    }
  }
}
```

**Comment by achiar99** [comment]
@babyhuey So i just copied your resource to my folder and run checkov and CKV_AWS_19 passed for this bucket
```Check: CKV_AWS_19: "Ensure all data stored in the S3 bucket is securely encrypted at rest"
        PASSED for resource: aws_s3_bucket.zappa-uap-us-west-1
        File: /aa.tf:1-4```
try to upgrade your checkov version to latest and re-run

**Comment by achiar99** [comment]
@okamosy can you please share the resource that failed?

**Comment by okamosy** [comment]
yeah, they are. but they are part of a larger module. I'm working to create a pared-down test for this as well.

**Comment by babyhuey** [comment]
> @babyhuey So i have just copied your resource to my folder and run checkov and CKV_AWS_19 passed for this bucket
> 
> ```
> Check: CKV_AWS_19: "Ensure all data stored in the S3 bucket is securely encrypted at rest"
>         PASSED for resource: aws_s3_bucket.zappa-uap-us-west-1
>         File: /aa.tf:1-4
> ```
> 
> Try to upgrade your checkov version to latest and re-run `pip3 install checkov --upgrade`

Yea, I was out of date. I swear I updated yesterday, but CKV_AWS_19 is good now!

**Comment by MrHash** [comment]
I'm also getting `CKV_AWS_20` and `CKV_AWS_57` errors on this config.

```
resource "aws_s3_bucket" "cloudfront" {
  bucket        = "cloudfront-example"
}

resource "aws_s3_bucket_public_access_block" "cloudfront" {
  bucket                  = aws_s3_bucket.cloudfront.bucket
  block_public_acls       = true
  block_public_policy     = true
  ignore_public_acls      = true
  restrict_public_buckets = true
}
```

**Comment by gruebel** [comment]
> ```yaml
> resource "aws_s3_bucket" "bucket" {
>   bucket = var.bucket_name
> 
>   tags = merge(
>     var.tags,
>     {
>       "Name" = var.bucket_name
>     }
>   )
> }
> 
> resource "aws_s3_bucket_server_side_encryption_configuration" "encryption" {
>   bucket = aws_s3_bucket.bucket.id
> 
>   rule {
>     apply_server_side_encryption_by_default {
>       sse_algorithm = "AES256"
>     }
>   }
> }
> ```

@okamosy I tested it with the given resources and the check passes, not sure what else you have there, which is somehow interferes, but from my testing I can say it works
```shell
Check: CKV_AWS_19: "Ensure all data stored in the S3 bucket is securely encrypted at rest"
        PASSED for resource: aws_s3_bucket.bucket
        File: /../tf/tf_s3.json:14-22
        Guide: https://docs.bridgecrew.io/docs/s3_14-data-encrypted-at-rest
```

**Comment by elvis2** [comment]
Howdy. This issue has been open a while. When will the v4 changes be added?

I ran into a S3 version false positive:
https://docs.bridgecrew.io/docs/s3_16-enable-versioning

I'm using the AWS 4.x provider. The example in your documentation is referencing Terraform attributes for AWS v3.
